### PR TITLE
quantize color images to reference lookup table

### PR DESCRIPTION
**Pull request by dyollb**
added option to quantize color images to reference lookup table (lut)
improved performance of lut IO (8s > 0.001s for VKF). this is a new format, which requires an update in S4L

### CODE REVIEW

File: Core/ColorLookupTable.cpp [hunk 1]
```diff
@@ -41,6 +41,16 @@ void ColorLookupTable::SetColor(size_t idx, unsigned char rgb[3])
 						 1.0);
 }
 
+void ColorLookupTable::GetColor(size_t idx, unsigned char uchar_rgb[3]) const
```
Comment (sanderegg) [inline Core/ColorLookupTable.cpp]: would be nicer with a color object

File: Core/ImageReader.cpp [hunk 2]
```diff
@@ -59,7 +59,7 @@ bool ImageReader::getInfo2D(const char* filename, unsigned& width, unsigned& hei
 	return ImageReader::getInfo(filename, width, height, nrslices, spacing, tr);
 }
 
-bool ImageReader::getImage2D(const char* filename, float* img, unsigned width, unsigned height, const std::function<float(unsigned char, unsigned char, unsigned char)>& color2grey)
+bool ImageReader::getImageStack(const std::vector<const char*>& filenames, float** img_stack, unsigned width, unsigned height, const std::function<float(unsigned char, unsigned char, unsigned char)>& color2grey)
```
Comment (sanderegg) [inline Core/ImageReader.cpp]: std::vector<std::string> instead of vector<const char*>

Comment (sanderegg) [reply Core/ImageReader.cpp]: returning a vector of smart unique pointer of images instead of a pointer of pointer of floats seems more secure

Comment (sanderegg) [reply Core/ImageReader.cpp]: what comes out when you just define __unsigned__ as a type? is this automatically int?

File: iSeg/LoaderWidgets.cpp [hunk 3]
```diff
@@ -1254,80 +979,169 @@ LoaderColorImages::LoaderColorImages(SlicesHandler* hand3D, eImageType typ, std:
...
+	xlength = new QSpinBox(0, 2000, 1, nullptr);
+	xlength->setValue(256);
 
-	vbox1->setFixedSize(vbox1->sizeHint());
+	auto yl = new QLabel(QString("y-Length: "));
+	ylength = new QSpinBox(0, 2000, 1, nullptr);
+	ylength->setValue(256);
 
-	subsect_toggled(subsect->isChecked());
+	auto subsect_layout = new QGridLayout(2,4);
```
Comment (sanderegg) [inline iSeg/LoaderWidgets.cpp]: you can also pass the layout in the widget constructor (the layout acts as parent if I'm not wrong). or use the qt designer which moves all this GUI stuff in XML-based files.

File: iSeg/LoaderWidgets.cpp [hunk 4]
```diff
@@ -1254,80 +979,169 @@ LoaderColorImages::LoaderColorImages(SlicesHandler* hand3D, eImageType typ, std:
...
 	}
 }
 
-void LoaderColorImages::subsect_toggled()
+void LoaderColorImages::load_quantize()
 {
-	bool isset = subsect->isOn();
-	if (isset)
+	QString filename = QFileDialog::getOpenFileName(QString::null,
+		"iSEG Color Lookup Table (*.lut *.h5)\nAll(*.*)", this);
```
Comment (sanderegg) [inline iSeg/LoaderWidgets.cpp]: If you want translation later on. There is a tr() function (http://doc.qt.io/qt-5/i18n-source-translation.html)

File: iSeg/SlicesHandler.cpp [hunk 5]
```diff
@@ -3908,6 +3522,14 @@ void SlicesHandler::clear_overlay()
 		_overlay[i] = 0.0f;
 }
 
+void SlicesHandler::new_overlay()
+{
+	if (_overlay != nullptr)
+		free(_overlay);
```
Comment (sanderegg) [inline iSeg/SlicesHandler.cpp]: this does not look very C++... free and malloc...

File: Core/ImageReader.cpp (hunk 2 above)
Comment (dyollb) [reply Core/ImageReader.cpp]: unsigned = unsigned int

Comment (dyollb) [reply Core/ImageReader.cpp]: I am passing the memory, which was already allocated, as vector of pointers -> it is an input.

Comment (dyollb) [reply Core/ImageReader.cpp]: I agree with all your comments, but did not want to change the API already used all over the place.

File: iSeg/LoaderWidgets.cpp (hunk 4 above)
Comment (dyollb) [reply iSeg/LoaderWidgets.cpp]: ok, thanks

File: iSeg/SlicesHandler.cpp (hunk 5 above)
Comment (dyollb) [reply iSeg/SlicesHandler.cpp]: legacy code, but I cannot delete something that was malloc'd somewhere else.
maybe next refactoring I will replace the overlay by a std::vector<float>

**Pull request merged**